apps/telegram_bot/main.py

```diff
@@ -0,0 +1,335 @@
+    def __init__(self):
+        # Initialize an empty dictionary to store user data
+        self.db = {}
+
+    def __getitem__(self, chat_id):
+        # Return the user data for a given user id, or create a new one if not found
+        if chat_id not in self.db:
+            self.db[chat_id] = {
+                "voices": DEFAULT_VOICES,
+                "current_voice": DEFAULT_VOICES[0],
+                "current_conversation": None,
+            }
+        return self.db[chat_id]
+
+    def __setitem__(self, chat_id, user_data):
+        # Set the user data for a given user id
+        self.db[chat_id] = user_data
+
+
+class VocodeBotResponder:
+    def __init__(self, transcriber, system_prompt, synthesizer, db=None):
+        self.transcriber = transcriber
+        self.system_prompt = system_prompt
+        self.synthesizer = synthesizer
+        self.db = db if db else InMemoryDB()
+        self.chat_ids = []
+
+    def get_agent(self, chat_id):
+        # Get current voice name and description from DB
+        _, voice_name, voice_description = self.db[chat_id].get(
+            "current_voice", (None, None, None)
+        )
+
+        # Augment prompt based on available info
+        prompt = self.system_prompt
+        if voice_description != None:
+            prompt += "The user described your voice as '{0}''. This is a demo of Coqui TTS's voice creation tool, so your responses are fun and relevant to that voice description."
+        if voice_name != None:
+            prompt += (
+                "You are {prompt}. Act like {prompt} and identify yourself as {prompt}."
+            )
+
+        agent = ChatGPTAgent(
+            system_prompt=self.system_prompt.format(voice_name),
+        )
+        # Load saved conversation if it exists
+        if self.db[chat_id]["current_conversation"]:
+            convo_string = self.db[chat_id]["current_conversation"]
+            agent.memory = pickle.loads(convo_string)
+
+        return agent
+
+    # input can be audio segment or text
+    async def get_response(
+        self, chat_id, input: Union[str, AudioSegment]
+    ) -> Tuple[str, AudioSegment]:
+        # If input is audio, transcribe it
+        if isinstance(input, AudioSegment):
+            input = self.transcriber.transcribe(input)
+
+        # Get agent response
+        agent_response = self.get_agent(chat_id).respond(input)
+
+        # Set current synthesizer voice from db
+        voice_id, _, voice_description = self.db[chat_id].get(
+            "current_voice", (None, None, None)
+        )
+
+        # If we have a Coqui voice prompt, use that. Otherwise, set ID as synthesizer expects.
+        if voice_description is not None:
+            self.synthesizer.voice_prompt = voice_description
+        else:
+            setattr(self.synthesizer, voice_attr_of[type(self.synthesizer)], voice_id)
+
+        # Synthesize response
+        synth_response = await self.synthesizer.async_synthesize(agent_response)
```

Comment:
this will fail for other synthesizers right? should we make a parameter to decide which method to use?